warehouse/macaroons/auth_policy.py

```diff
@@ -0,0 +1,147 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+
+from pyramid.authentication import CallbackAuthenticationPolicy
+from pyramid.interfaces import IAuthenticationPolicy, IAuthorizationPolicy
+from pyramid.security import Denied
+from pyramid.threadlocal import get_current_request
+from zope.interface import implementer
+
+from warehouse.cache.http import add_vary_callback
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.macaroons.services import InvalidMacaroon
+
+
+def _extract_basic_macaroon(auth):
+    """
+    A helper function for extracting a macaroon from a
+    HTTP Basic Authentication-style header.
+
+    Returns None if the header doesn't contain a structurally
+    valid macaroon, or the candidate (not yet verified) macaroon
+    in a serialized form.
+    """
+    try:
+        authorization = base64.b64decode(auth).decode()
+        auth_method, _, auth = authorization.partition(":")
+    except ValueError:
+        return None
+
+    if auth_method != "@token":
+        return None
+
+    return auth
+
+
+def _extract_http_macaroon(request):
+    """
+    A helper function for the extraction of HTTP Macaroon from a given request.
+    Returns either a None if no macaroon could be found, or the byte string
```

Comment:
not 100% sure, but this looks like a string, not byte string. We're splitting with `" "` and not `b" "`, and matching with `"basic"` and `"token"`, not their bytes counterparts.